tests/unit/s2n_client_hello_get_supported_groups_test.c

```diff
@@ -0,0 +1,400 @@
+/*
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include "pq-crypto/s2n_pq.h"
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/extensions/s2n_client_supported_groups.h"
+#include "tls/s2n_client_hello.h"
+#include "tls/s2n_tls.h"
+#include "utils/s2n_random.h"
+
+#define S2N_TEST_SUPPORTED_GROUPS_LIST_COUNT 16
+
+/* Each supported group is 2 bytes. */
+#define S2N_TEST_SUPPORTED_GROUPS_LIST_SIZE (S2N_TEST_SUPPORTED_GROUPS_LIST_COUNT * S2N_SUPPORTED_GROUP_SIZE)
+
+/* 2 length bytes + space for the list of supported groups. */
+#define S2N_TEST_SUPPORTED_GROUPS_EXTENSION_SIZE (2 + S2N_TEST_SUPPORTED_GROUPS_LIST_SIZE)
+
+struct s2n_client_hello_context {
+    const struct s2n_security_policy *client_security_policy;
+    unsigned client_supports_pq : 1;
+    int invoked_count;
+};
+
+int s2n_check_received_supported_groups_cb(struct s2n_connection *conn, void *ctx)
+{
+    EXPECT_NOT_NULL(ctx);
+
+    struct s2n_client_hello_context *context = (struct s2n_client_hello_context *) ctx;
+    context->invoked_count += 1;
+
+    struct s2n_client_hello *client_hello = s2n_connection_get_client_hello(conn);
+    EXPECT_NOT_NULL(client_hello);
+
+    uint16_t supported_groups[S2N_TEST_SUPPORTED_GROUPS_LIST_COUNT] = { 0 };
+    uint16_t supported_groups_count = 0;
+    EXPECT_SUCCESS(s2n_client_hello_get_supported_groups(client_hello, supported_groups,
+            s2n_array_len(supported_groups), &supported_groups_count));
+
+    const struct s2n_security_policy *security_policy = context->client_security_policy;
+    uint16_t expected_groups_count = security_policy->ecc_preferences->count;
+    if (context->client_supports_pq) {
+        expected_groups_count += security_policy->kem_preferences->tls13_kem_group_count;
+    }
+    EXPECT_EQUAL(supported_groups_count, expected_groups_count);
```

Comment:
This test might be a little brittle. I think groups are currently the only aspect of our security policy where this is reliably true-- for the others, there's more logic around whether or not to send each option in the list. And I think once we clean up the conditional compilation around kem groups (https://github.com/aws/s2n-tls/issues/4100) this might break. 

It's probably fine for now, but what would you do (/ what do you want a future dev to do) if you can no longer rely on these counts? Maybe just check >0 && < the count from the policy?

Reply:
I changed the test to compare the retrieved supported groups to whatever was written by the client. This way if what the client writes is changed the test won't break. Let me know if this seems okay.